File: create/tools/import_from_icloud_sr.go
package tools

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/nyamiiko/script_hub/go_scripts/hub"
)

var categoryMap = map[string]string{
	"amplify_nexus": "『 🛠️ Amplify Nexus › 增幅枢纽 』",
	"head_expanse":  "『 🔝 Head Expanse › 首端扩域 』",
}

func getModuleName(content string) string {
	m := regexp.MustCompile(`(?m)^#!name\s*[=:]\s*(.+)$`).FindStringSubmatch(content)
	if m != nil {
		return strings.TrimSpace(m[1])
	}
	return ""
}

func getContentHash(content string) string {
	var filtered []string
	for _, line := range strings.Split(content, "\n") {
		if !strings.HasPrefix(line, "#!category") && !strings.HasPrefix(line, "#!url") {
			filtered = append(filtered, line)
		}
	}
	hash := md5.Sum([]byte(strings.Join(filtered, "\n")))
	return hex.EncodeToString(hash[:])[:8]
}

func classifyModule(name, content string) string {
	nameLower := strings.ToLower(name)
	kwsNexus := []string{"wifi", "calling", "helper", "enhanced", "dns", "iringo", "dualsubs", "tiktok", "助手"}
	for _, k := range kwsNexus {
		if strings.Contains(nameLower, k) {
			return "amplify_nexus"
		}
	}
	kwsExpanse := []string{"ad-block", "adblock", "firewall", "script hub", "广告平台", "广告联盟", "universal"}
	for _, k := range kwsExpanse {
		if strings.Contains(nameLower, k) {
			return "head_expanse"
		}
	}
	return "narrow_pierce"
}

func RunImportFromIcloudSr() int {
	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("🚀 Shadowrocket to Surge Module Importer")
	fmt.Println(strings.Repeat("=", 60))

	srDirStr := os.Getenv("SHADOWROCKET_MODULES_DIR")
	if srDirStr == "" {
		homeDir, _ := os.UserHomeDir()
		srDirStr = filepath.Join(homeDir, "Library/Mobile Documents/iCloud~com~liguangming~Shadowrocket/Documents/Modules")
	}

	if !hub.ValidateDirExists(srDirStr, "") {
		fmt.Printf("❌ Shadowrocket directory not found: %s\n", srDirStr)
		return 1
	}

	surgeDir := filepath.Join(hub.ROOT, "modules/surge")

	type existingInfo struct {
		path string
		hash string
		name string
	}
	existing := make(map[string]existingInfo)

	for _, cat := range []string{"amplify_nexus", "head_expanse"} {
		catPath := filepath.Join(surgeDir, cat)
		if !hub.ValidateDirExists(catPath, "") {
			continue
		}
		entries, _ := os.ReadDir(catPath)
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".sgmodule") {
				fPath := filepath.Join(catPath, e.Name())
				content := hub.ReadFileString(fPath)
				if content == "" {
					continue
				}
				name := getModuleName(content)
				if name == "" {
					name = strings.TrimSuffix(e.Name(), ".sgmodule")
				}
				existing[strings.ToLower(name)] = existingInfo{
					path: fPath,
					hash: getContentHash(content),
					name: name,
				}
			}
		}
	}

	fmt.Printf("Existing modules found: %d\n\n", len(existing))

	added, duplicate, skipped := 0, 0, 0

	entries, err := os.ReadDir(srDirStr)
	if err != nil {
		fmt.Println("❌ Failed to read SR dir")
		return 1
	}

	for _, e := range entries {
		if e.IsDir() || strings.HasPrefix(e.Name(), "__") || !strings.Contains(e.Name(), "module") {
			continue
		}
		srFile := filepath.Join(srDirStr, e.Name())
		info, err := e.Info()
		if err != nil {
			continue
		}
		if info.Size() > 100000 {
			fmt.Printf("⏭️  Skipping large file: %s (%dKB)\n", e.Name(), info.Size()/1024)
			skipped++
			continue
		}

		content := hub.ReadFileString(srFile)
		if content == "" {
			fmt.Printf("❌ Failed to read: %s\n", e.Name())
			continue
		}

		moduleName := getModuleName(content)
		if moduleName == "" {
			decoded, _ := url.QueryUnescape(strings.TrimSuffix(e.Name(), filepath.Ext(e.Name())))
			moduleName = decoded
		}
		contentHash := getContentHash(content)
		nameKey := strings.ToLower(moduleName)

		if ex, found := existing[nameKey]; found {
			if ex.hash == contentHash {
				fmt.Printf("🔄 Duplicate: %s\n", moduleName)
				duplicate++
			} else {
				fmt.Printf("📝 Exists with different content: %s\n", moduleName)
				fmt.Printf("   → Keeping existing: %s\n", filepath.Base(ex.path))
				skipped++
			}
			continue
		}

		category := classifyModule(moduleName, content)
		safeName := regexp.MustCompile(`[<>:"/\\|?*]`).ReplaceAllString(moduleName, "")
		if !strings.HasSuffix(safeName, ".sgmodule") {
			safeName += ".sgmodule"
		}
		dstPath := filepath.Join(surgeDir, category, safeName)

		lines := strings.Split(content, "\n")
		var newLines []string
		catAdded := false

		for _, line := range lines {
			if strings.HasPrefix(line, "#!url") {
				continue
			}
			if strings.HasPrefix(line, "#!category") {
				if !catAdded {
					newLines = append(newLines, fmt.Sprintf("#!category=%s", categoryMap[category]))
					catAdded = true
				}
				continue
			}
			if strings.HasPrefix(line, "#!name") && !catAdded {
				newLines = append(newLines, fmt.Sprintf("#!category=%s", categoryMap[category]))
				catAdded = true
			}
			newLines = append(newLines, line)
		}

		if !catAdded {
			newLines = append([]string{fmt.Sprintf("#!category=%s", categoryMap[category])}, newLines...)
		}

		hub.SafeWriteFile(dstPath, strings.Join(newLines, "\n"), true)
		fmt.Printf("✅ Added: %s → %s/\n", moduleName, category)
		added++
	}

	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Printf("Stats: Added %d, Duplicate %d, Skipped %d\n", added, duplicate, skipped)
	fmt.Println(strings.Repeat("=", 60))

	return 0
}

File: create/scripts/tools/import_from_icloud_sr.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Import modules from Shadowrocket to Surge (Incremental + Dedup)
Features:
1. Sync modules from iCloud Shadowrocket directory
2. Deduplicate based on content hash
3. Preserve original category and naming
"""

import os
import re
import hashlib
from pathlib import Path
from urllib.parse import unquote

# Project Paths
SCRIPT_DIR = Path(__file__).resolve().parent
PROJECT_ROOT = SCRIPT_DIR.parent.parent
SR_DIR = Path(
    os.environ.get(
        "SHADOWROCKET_MODULES_DIR",
        "~/Library/Mobile Documents/iCloud~com~liguangming~Shadowrocket/Documents/Modules",
    )
).expanduser()
SURGE_DIR = PROJECT_ROOT / "modules" / "surge"

# Category Mapping (Preserve original Chinese group names as per user instruction)
CATEGORY_MAP = {
    "amplify_nexus": "『 🛠️ Amplify Nexus › 增幅枢纽 』",
    "head_expanse": "『 🔝 Head Expanse › 首端扩域 』",
}

def get_module_name(content):
    """Extract #!name from module content."""
    match = re.search(r'^#!name\s*[=:]\s*(.+)$', content, re.MULTILINE)
    return match.group(1).strip() if match else None

def get_content_hash(content):
    """Calculate content hash (ignore metadata like category/url)."""
    lines = content.split('\n')
    filtered = [line_ for line_ in lines if not line_.startswith('#!category') and not line_.startswith('#!url')]
    return hashlib.md5('\n'.join(filtered).encode()).hexdigest()[:8]

def classify_module(name, content):
    """Classify module based on name keywords."""
    name_lower = name.lower()
    if any(k in name_lower for k in ['wifi', 'calling', 'helper', 'enhanced', 'dns', 'iringo', 'dualsubs', 'tiktok', '助手']):
        return "amplify_nexus"
    if any(k in name_lower for k in ['ad-block', 'adblock', 'firewall', 'script hub', '广告平台', '广告联盟', 'universal']):
        return "head_expanse"
    return "narrow_pierce"

def main():
    print("=" * 60)
    print("🚀 Shadowrocket to Surge Module Importer")
    print("=" * 60)

    if not SR_DIR.exists():
        print(f"❌ Shadowrocket directory not found: {SR_DIR}")
        return

    # Scan existing modules
    existing = {}  # name_lower -> {path, hash, name}
    for cat in ["amplify_nexus", "head_expanse"]:
        cat_path = SURGE_DIR / cat
        if not cat_path.exists():
            continue
        for f in cat_path.glob("*.sgmodule"):
            try:
                content = f.read_text(encoding='utf-8')
                name = get_module_name(content) or f.stem
                existing[name.lower()] = {
                    "path": f,
                    "hash": get_content_hash(content),
                    "name": name
                }
            except Exception:
                pass

    print(f"Existing modules found: {len(existing)}\n")

    added = duplicate = skipped = 0

    # Process SR modules
    for sr_file in sorted(SR_DIR.glob("*.*module")):
        filename = sr_file.name
        if filename.startswith("__"):
            continue

        size = sr_file.stat().st_size
        if size > 100000:
            print(f"⏭️  Skipping large file: {filename} ({size//1024}KB)")
            skipped += 1
            continue

        try:
            content = sr_file.read_text(encoding='utf-8')
        except Exception:
            print(f"❌ Failed to read: {filename}")
            continue

        module_name = get_module_name(content) or unquote(sr_file.stem)
        content_hash = get_content_hash(content)
        name_key = module_name.lower()

        if name_key in existing:
            ex = existing[name_key]
            if ex["hash"] == content_hash:
                print(f"🔄 Duplicate: {module_name}")
                duplicate += 1
            else:
                print(f"📝 Exists with different content: {module_name}")
                print(f"   → Keeping existing: {ex['path'].name}")
                skipped += 1
            continue

        category = classify_module(module_name, content)
        safe_name = re.sub(r'[<>:"/\\|?*]', '', module_name)
        if not safe_name.endswith('.sgmodule'):
            safe_name += '.sgmodule'
        dst_path = SURGE_DIR / category / safe_name

        # Process content
        lines = content.split('\n')
        new_lines = []
        cat_added = False
        for line in lines:
            if line.startswith('#!url'):
                continue
            if line.startswith('#!category'):
                if not cat_added:
                    new_lines.append(f"#!category={CATEGORY_MAP[category]}")
                    cat_added = True
                continue
            if line.startswith('#!name') and not cat_added:
                new_lines.append(f"#!category={CATEGORY_MAP[category]}")
                cat_added = True
            new_lines.append(line)

        if not cat_added:
            new_lines.insert(0, f"#!category={CATEGORY_MAP[category]}")

        dst_path.write_text('\n'.join(new_lines), encoding='utf-8')
        print(f"✅ Added: {module_name} → {category}/")
        added += 1

    print("\n" + "=" * 60)
    print(f"Stats: Added {added}, Duplicate {duplicate}, Skipped {skipped}")
    print("=" * 60)

if __name__ == "__main__":
    main()
